rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala

```diff
@@ -65,14 +65,14 @@ object SpatialMatcher {
       .reverse
 
   // This helper function is useful in several productions
-  def foldMatch(tlist: Seq[Par], plist: Seq[Par]): OptionalFreeMap[Unit] =
+  def foldMatch[T](tlist: Seq[T],
+                   plist: Seq[T],
+                   f: (T, T) => OptionalFreeMap[Unit]): OptionalFreeMap[Unit] =
```

Comment:
Why is this necessary? Does scala not know how to dispatch to the corresponding methods

spatialMatch(target: Par, pattern: Par)
spatialMatch(target: Channel, pattern: Channel)

based on the types passed in?

Reply:
`T` is too generic for that.  There isn't a `spatialMatch` for any old type, so taking that approach, you'd have to constrain `T` with a type class (something like `trait SpatialMatch[A]`).  Then you'd have in something like "any `T` that has an instance of `SpatialMatch`" (aka something of type `SpatialMatch[T]`).

Or alternately, if `Par` or `Channel` were both subtypes of another type (or even if one was a subtype of the other), we could use the supertype and we could use `spatialMatch` in the body of the function.

Reply:
I see. I guess in the future everything as in (Par, Expr, Send, Receive, etc) should extend from a trait Term and then have a generic spatialMatch(target: Term, pattern: Term). But yeah I guess the other spatialMatch methods need to be implemented first.

Reply:
I'm fine with this change and the associated changes here.